### PR TITLE
use contextual components

### PR DESCRIPTION
**Pull request by null-null-null**
@firefly-ui/owners @pzuraq 

Convert to contextual components. Lets us remove the use of the private `nearestOfType`, while also providing a cleaner public API.

### CODE REVIEW

File: app/utils/sort-by-dom-position.js [hunk 1]
```diff
@@ -0,0 +1 @@
+export { default } from 'ui-option-list/utils/sort-by-dom-position';
```
Comment (twokul) [inline app/utils/sort-by-dom-position.js]: no newline at the end of file

File: addon/components/ui-option-list.js [hunk 2]
```diff
@@ -35,39 +36,69 @@ export default Ember.Component.extend({
...
+      } else {
+        if (this.attrs['on-keyboard-out']) {
+          this.attrs['on-keyboard-out']();
+        } else {
+          options[options.length - 1].focus();
+        }
+      }
+    },
+
+    register(option) {
```
Comment (pzuraq) [inline addon/components/ui-option-list.js]: This is a pattern I definitely want to stay away from, it's a source of headaches when Options are managed so directly by Option Lists. Lot's of bugs, codependency, etc.

File: addon/components/ui-option-list.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import Ember from 'ember';
 import layout from '../templates/components/ui-option-list';
+import sortByDomPosition from '../utils/sort-by-dom-position';
```
Comment (pzuraq) [inline addon/components/ui-option-list.js]: Don't like the `/utils/...` pattern. It becomes a bucket of miscellaneous helper functions, source of spaghetti. Let's consider keeping helper functions like this local until they need to be abstracted.

File: addon/components/ui-option-list.js [hunk 4]
```diff
@@ -35,39 +36,69 @@ export default Ember.Component.extend({
...
+    deselectOption(value) {
+      this.sendAction('on-deselect', value);
+    },
+
+    handleUnregisteredKeyEvent(event) {
+      this.sendAction('on-key-down', event);
+    },
+
+    moveDown(option) {
+      const options = sortByDomPosition(this.get('options').map((option) => option.element));
```
Comment (pzuraq) [inline addon/components/ui-option-list.js]: Seems inefficient to do this every time we move up or down.

File: addon/components/ui-option-list.js (hunk 2 above)
Comment (null-null-null) [inline addon/components/ui-option-list.js]: Does beg the question of whose responsibility this is. Previously, moveUp/moveDown was handled by the options themselves, which meant they had to be aware of their peers, and in turn their parent (to find their peers in the first place).

Comment (pzuraq) [inline addon/components/ui-option-list.js]: Yes, but they were only aware at a DOM context level. The keyboard navigation behavior, while important, was sort of secondary and so I decided to go the route of treating it like I was just using plain jQuery. That way the components themselves wouldn't become _too_ intertwined.

This is why I really want to completely redo these components though. I feel like we need to go back to the drawing board and ask ourselves what the responsibilities are, how they should work with each other, etc. Fresh start.

Comment (null-null-null) [inline addon/components/ui-option-list.js]: Is the plan to focus on Firefly UI next sprint? I definitely want to be a part of those conversations. Until then, I'm happy to close this PR.

**Pull request closed without merging**